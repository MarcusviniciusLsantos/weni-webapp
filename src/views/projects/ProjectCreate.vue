<template>
  <ContainerCondensed>
    <header>
      <h6 class="u color-neutral-darkest">
        {{ $t('projects.create.title') }}
      </h6>

      <div class="u font secondary body-lg color-neutral-cloudy">
        {{ $t('projects.create.subtitle') }}
      </div>
    </header>

    <UnnnicInputNext
      :value="projectName"
      @input="
        projectName = $event;
        errors.projectName = false;
      "
      :label="$t('orgs.create.project_name')"
      :placeholder="$t('orgs.create.project_name_placeholder')"
      :error="errors.projectName"
      ref="projectName"
    />

    <DescriptionTextarea
      ref="projectDescription"
      class="mt-sm"
      :value="projectDescription"
      :error="errors.projectDescription"
      @input="
        projectDescription = $event;
        errors.projectDescription = false;
      "
    />

    <UnnnicSelect
      v-model="dateFormat"
      :label="$t('orgs.create.date_format')"
    >
      <option value="D">DD-MM-YYYY</option>
      <option value="M">MM-DD-YYYY</option>
    </UnnnicSelect>

    <ProjectFormatControl
      :type="projectFormat"
      @change="
        projectFormat = $event;
        errors.projectFormat = false;
      "
      :setup.sync="setupFields"
      :error="errors.projectFormat"
      ref="projectFormat"
    />

    <div class="weni-create-org__group weni-create-org__group__buttons">
      <UnnnicButton
        type="tertiary"
        :disabled="loadingButton"
        @click="onBack()"
      >
        {{ $t('orgs.create.back') }}
      </UnnnicButton>
      <UnnnicButton
        :loading="loadingButton"
        type="secondary"
        @click="onCreateProject()"
      >
        {{ $t('projects.create.create') }}
      </UnnnicButton>
    </div>
  </ContainerCondensed>
</template>

<script>
import { mapActions, mapGetters, mapState } from 'vuex';
import timezones from './timezone';
import ProjectFormatControl from './ProjectFormatControl.vue';
import ContainerCondensed from '../../components/ContainerCondensed.vue';
import DescriptionTextarea from './form/DescriptionTextarea.vue';

export default {
  name: 'ProjectCreate',
  components: {
    ContainerCondensed,
    ProjectFormatControl,
    DescriptionTextarea,
  },

  mixins: [timezones],

  data() {
    return {
      projectName: null,
      projectDescription: '',
      dateFormat: 'D',
      timeZone: 'America/Sao_Paulo',
      loading: false,
      project: null,
      projectFormat: null,
      setupFields: {},
      errors: {
        projectName: false,
        projectDescription: false,
        projectFormat: false,
      },
    };
  },
  computed: {
    ...mapState({
      loadingButton: (state) => state.Project.loadingCreateProject,
    }),
    ...mapGetters(['currentOrg', 'currentProject']),
  },
  methods: {
    ...mapActions(['createProjectForOrg', 'setCurrentProject', 'openModal']),

    onBack() {
      this.$router.push({
        name: 'projects',
        params: {
          orgUuid: this.currentOrg.uuid,
        },
      });
    },
    onAccess(uuid) {
      this.$router.push({
        name: 'home',
        params: {
          projectUuid: uuid,
        },
      });
      this.$root.$emit('set-sidebar-expanded');
    },
    async onCreateProject() {
      const canFinish = [
        this.projectName,
        this.projectDescription,
        this.dateFormat,
        this.timeZone,
        this.projectFormat,
      ].every((field) => field && field.length > 0);

      if (!canFinish) {
        ['projectFormat', 'projectDescription', 'projectName'].forEach(
          (fieldName) => {
            if (!this[fieldName]) {
              console.log('find', fieldName, this.$refs[fieldName]);
              this.$refs[fieldName].$el.scrollIntoView({
                behavior: 'smooth',
                inline: 'nearest',
              });
              this.errors[fieldName] = this.$t('errors.required');
            }
          },
        );

        return;
      }

      await this.createProjectForOrg({
        project: {
          name: this.projectName,
          description: this.projectDescription,
          dateFormat: this.dateFormat,
          timeZone: this.timeZone,
          format: this.projectFormat,
          globals: this.setupFields,
        },
        onBack: this.onBack,
        onAccess: this.onAccess,
      });
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~@weni/unnnic-system/src/assets/scss/unnnic.scss';

.mt-sm {
  margin-top: $unnnic-spacing-sm;
}

header {
  text-align: center;
  margin-bottom: $unnnic-spacing-stack-md;
  display: flex;
  flex-direction: column;
  row-gap: $unnnic-spacing-stack-xs;
}

.unnnic-select ::v-deep .unnnic-form__label {
  margin-top: $unnnic-spacing-stack-md;
}
</style>
