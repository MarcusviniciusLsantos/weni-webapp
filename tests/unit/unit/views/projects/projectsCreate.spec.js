import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import ProjectCreate from '@/views/projects/ProjectCreate.vue';
import i18n from '@/utils/plugins/i18n';
import { org } from '../../../__mocks__/';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('ProjectCreate.vue', () => {
  let getters;
  let state;
  let actions;
  let store;
  let wrapper;

  beforeEach(() => {
    state = {
      Project: {
        loadingCreateProject: false,
      },
    };
    getters = {
      currentOrg: jest.fn(() => org),
    };
    actions = {
      setCurrentProject: jest.fn(),
      createProject: jest.fn(),
      createProjectForOrg: jest.fn(),
    };

    store = new Vuex.Store({
      getters,
      actions,
      state,
    });

    wrapper = shallowMount(ProjectCreate, {
      localVue,
      i18n,
      store,
      stubs: {
        UnnnicInput: true,
        UnnnicSelect: true,
        UnnnicButton: true,
        UnnnicInputNext: true,
        Container: true,
      },
      mocks: {
        $t: () => 'some specific text',
        $router: {
          push: jest.fn(),
        },
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders a snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('sets the correct default data', () => {
    expect(typeof ProjectCreate.data).toBe('function');
    const defaultData = ProjectCreate.data();
    expect(defaultData.projectName).toBe(null);
    expect(defaultData.dateFormat).toBe('D');
    expect(defaultData.timeZone).toBe('America/Sao_Paulo');
    expect(defaultData.loading).toBe(false);
    expect(defaultData.project).toBe(null);
  });

  it('should change route in onBack()', () => {
    const spy = jest.spyOn(wrapper.vm.$router, 'push');
    wrapper.vm.onBack();

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith({
      name: 'projects',
      params: { orgUuid: getters.currentOrg().uuid },
    });
  });

  it('should be able to test onAccess function', async () => {
    wrapper.setData({
      project: {
        uuid: 'uuid',
        organization: {
          uuid: getters.currentOrg.uuid,
        },
        name: 'project name',
        flow_organization: {
          uuid: 'flow org uuid',
        },
        menu: 'project menu',
      },
    });

    expect(wrapper.vm.project).toBeDefined();

    const spy = jest.spyOn(wrapper.vm.$router, 'push');
    wrapper.vm.onAccess('uuid');

    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith({
      name: 'home',
      params: { projectUuid: 'uuid' },
    });
  });
});
