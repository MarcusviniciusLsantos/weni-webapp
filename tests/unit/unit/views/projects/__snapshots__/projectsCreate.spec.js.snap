// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCreate.vue renders a snapshot 1`] = `
<containercondensed-stub>
  <header>
    <h6 class="u color-neutral-darkest">
      some specific text
    </h6>
    <div class="u font secondary body-lg color-neutral-cloudy">
      some specific text
    </div>
  </header>
  <unnnicinputnext-stub label="some specific text" placeholder="some specific text"></unnnicinputnext-stub>
  <descriptiontextarea-stub value="" infomaxwidth="21.125rem" class="mt-sm"></descriptiontextarea-stub>
  <unnnicformelement label="some specific text">
    <unnnicselectsmart value="[object Object]" options="[object Object],[object Object]" autocomplete="" autocompleteclearonfocus=""></unnnicselectsmart>
  </unnnicformelement>
  <projectformatcontrol-stub setup="[object Object]"></projectformatcontrol-stub>
  <div class="weni-create-org__group weni-create-org__group__buttons">
    <unnnicbutton-stub type="tertiary">
      some specific text
    </unnnicbutton-stub>
    <unnnicbutton-stub type="secondary">
      some specific text
    </unnnicbutton-stub>
  </div>
</containercondensed-stub>
`;
